832 use jinja for wrapper

hiker opened the pull request:
This introduces jinja for creating the wrapper libraries, and base classes to improve code reusage. It needs PR #812 to be merged first. It actually does NOT fully close #832, since in order to keep this PR smaller I have not included the profiling wrapper (a separate branch for that ... 832_2... is work in progress).

hiker commented:
This PR introduces jinja to produce the read-only libraries (including a common base class). There are two decision that we might want to think about - I picked one solution in each case, but am not married to any of them:
1) As discussed in the Telco: we originally said that we would include all jinja-processed files into the repo (so compilation could be done without need to install jinja). During my implementation I decided that each library (e.g. readonly-dlsm, readonly-lfric) should pre-process and compile its own version of the PSyData- and ReadOnly base classes: first of all, so that these files would be as small as possible and not include unnecessary (and unused) functions (though it wouldn't hurt if they were there). Secondly, they should be compiled per library since we likely use different compilers for different libraries (e.g. GPU profiling vs other library), so we could not use one compiled library anyway. But it would be equally easy and possible to just have one jinja-processed source file (which would include all data types, 32/64-bit - real/int - and 0 (scalar), 1, 2, 3, 4 dimensions), which is then just compiled per library that uses it. Adding jinja as dependency is trivial, it's just a pip install.
2) After 1) I decided that the PSyDataBase class should provide the static functions (start, stop, init, finalise) with the proper name for the corresponding library (e.g. profile_PSyDataStart, read_only_verification_PSyDataStart, ...), since many libraries do not need to change these functions at all. If you want to provide your own version, you would have to rename these functions when 'use'-ing them in your library (``use psy_data_base_mod, only: base_PSyDataStart=>profile_PSyDataStart``), then you can provide your own version of ``profile_PSyDataStart`` and call the original implementation (now called ``base_PSyDataStart``). I picked this way because:
    1) mostly you don't need to do any of this, so a library can just use: ``use psy_data_base_mod, only: profile_PSyDataStart`` 
    and the function is then available for the application. The jinja-processing in the api-specific library will make sure that 
    the right prefix is used for the static functions.
    2) Often if you provide your own functions, there is no need to call the 'base class static function` anyway (e.g. 
    nvidia start/stop does not use the functionality provided at all, so no need to even use it, and it doesn't matter what it is 
    called in the first place - base_PSyDatStart, or profile_PSyDataStart - the function is just not used in the wrapper library)

    But similarly we could say that the static functions in PSyDataBase should be called (say) ``base_PSyDataStart``, then a wrapper library that does not want to overwrite these functions would rename them (``use psy_data_base_mod, only: profile_PSyDataStart=>base_PSyDataStart``) to make them available for the application with the proper prefix, and if you want to provide your own function, you can then call ``base_PSyDataStart`` to get its functionality if required.

I don't prefer one solution about the other :) Note that the profiling libraries will get a separate pull request where I use the jinja wrapper base class for some of them (mostly to use the verbosity functionality).

arporter commented on lib/read_only/read_only_base.jinja:
```diff
@@ -0,0 +1,365 @@
+{{ "! THIS FILE IS CREATED FROM THE JINJA TEMPLATE FILE " }}
+{{ "! DO NOT MODIFY DIRECTLY                            " }}
+{{ "! ==================================================" }}
+
```
Please add a comment here giving an overview of this file. e.g. "Jinja-templated implementation of XXX. Jinja templates are used to YYY"

hiker replied:
Done.

arporter replied:
Great, thanks.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+This will create the processed file ``psy_data_base.f90`` in the directory
+of the library (which has the advantage that you will be using consistent
+compiler settings in your library, and consistent parameter specifying
+the required types and dimensions). You still need to declare all these
```
Perhaps add a reference to either the lfric or gocean read-only implementation as an example (if that makes sense)?

hiker replied:
I moved that paragraph to a better place, and documented it a bit better. But yes, you somehow have to list all the automatically created functions in one generic interface. This can be done either 'manually' (e.g. dl_esm_inf readonly), or via a jinja script. I've improved the docs to point these two approaches out.
I also believe that just above the paragraph you pointed I already state where the example code comes from, so I don't think we need another reference here (since this paragraph actually explains the very same code).

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+PSyclone provides a base class for all PSyData wrapper libraries. The
+base class is independent of the API, but it can provide implementations for
+scalars and arrays for all native Fortran types that are required by the
+API-specific implementation. The base class needs not to be used, but it
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    It will check the ``PSYDATA_VERBOSE`` environment flag. If it exists, it
+    must have a value of either 0 (no messages), 1 (some messages, typically
+    only ``PSyDataStart`` and ``PSyDataEnd``), or 2 (detailed messages,
+    depending on wrapper library). All other values will return in a warning
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    must have a value of either 0 (no messages), 1 (some messages, typically
+    only ``PSyDataStart`` and ``PSyDataEnd``), or 2 (detailed messages,
+    depending on wrapper library). All other values will return in a warning
+    message to be printed (and verbosity will be disabled). The verbosity level
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    is available as ``this%verbosity``.
+
+Module- and Region-Name Handling:
+    The modules stores the module name in ``this%module_name``, and the region
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    It automatically sets ``this%next_var_index`` to 1 in ``PSyDataPreStart``
+    ``PSyDataPreEndDeclaration`` and ``PostStart``. This variable will also
+    be increased by one for each call to a Declare- or Provide-subroutine.
+    This variable can be used to have a reproducible index for declaring and
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+
+Start/Stop Handling:
+    The base class maintains a module variable ``is_enabled``. This is set
+    to true at startup, and gets enabled and disabled by calling the module
```
Agreed, I've opened #889.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+
+-types:
+    A comma-separated list of Fortran basic types (no spaces allowed).
+    Supported are the type names:
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    Default value is ``real,double,int``.
+
+-dims:
+    A comma-separated list of dimensions (no spaced allowed). Default
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+``DeclareArray{{dim}}d{{type}}`` and ``ProvideArray{{dim}}d{{type}}``
+will be created for each type and each specified number of dimensions.
+
+Below an example of using the ``process.py`` script in a Makefile for a read-only
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+
+This will create the processed file ``psy_data_base.f90`` in the directory
+of the library (which has the advantage that you will be using consistent
+compiler settings in your library, and consistent parameter specifying
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+the read-only-verification base class (see
+:ref:`psydata_read_only_base_class`) uses it. If more types are required,
+they can be defined in ``process.py``. If the additional types need
+different number of bits and are required in a read-only library, the
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+:ref:`psydata_read_only_base_class`) uses it. If more types are required,
+they can be defined in ``process.py``. If the additional types need
+different number of bits and are required in a read-only library, the
+read-only-verification base class needs to be adjusted as well.
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+different number of bits and are required in a read-only library, the
+read-only-verification base class needs to be adjusted as well.
+
+.. code-block:: jinja
```
There is some description there now.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+        ProvideFieldVectorDouble, &
+        {{all_provides|join(", &\n"+indent) }}
+
+.. note:
```
Gee - sphinx should really warn if this happens (i.e. I missed a ':').

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+        {{all_provides|join(", &\n"+indent) }}
+
+.. note:
+    Using ``or ""`` in the jinja statements avoids that ``None`` is
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+
+Jinja Support in the Base Class
++++++++++++++++++++++++++++++++
+The base class is contained in ``lib/psy_data_base.jinja``. It is processed
```
I've added this. I've considered two options - a separate introduction at the beginning of the description of the base class (i.e. before I mention which features are provided by the base class), or an intro in the subsection as you mentioned, and while it is a bit strange that we now first mention that jinja is a feature before we explain why, I think it actually works.

arporter replied:
Yes, I think it's good now - thanks.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+The ``FieldDouble`` and ``FieldVectorDouble`` related functions are explicitly
+coded in the library, the rest is taken from the base class:
+
+.. code-block:: jinja
```
I tried to clarify that this creates the definition of the generic interface.

arporter replied:
It's good now - thanks!

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    subroutine profile_PSyDataStart()
+      use psydata_base_mod, only : base_PSyDataStart => profile_PSyDataStart
+      ! Do something
+      call basePSyDataStart()
```
Indeed.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+      ! Do something else
+    end subroutine profile_PSyDataStart
+
+.. note:
```
My copy&paste error

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    end subroutine profile_PSyDataStart
+
+.. note:
+    The ``PSyDataBase`` class will use the prefix that you have specified
```
I've extended this a bit, hope it's better. The basic idea is that if you need to provide your own version of one of the static functions (but still need to use the functionality of the original class), you have to rename the original function somehow - either you create it using a different name, or you rename it when importing ('use'). I've opted for renaming, but thought it would be useful that you could also just change all prefixes in the base class using jinja.

arporter replied:
Oh yes, I get it now :-)

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+PSyData Read-Only-Verification Base Class
+-----------------------------------------
+The ReadOnlyVerification transformation uses the PSyData API
+to compute and store a checksum of each read-only parameter
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    The ``PreStart`` function gets the number of variables as
+    parameters. The decision not to use this value for allocating
+    the array is that the LFRic read-only implementation stores
+    several checksum for one variable of a certain type (VectorField).
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+allocated.
+
+.. note::
+    The ``PreStart`` function gets the number of variables as
```
Done.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    The ``DeclareVariable`` functions for VectorFields counts the
+    right number of checksums required.
+
+This base class uses the information about the number of bits for
```
Clarified.

hiker commented on doc/developer_guide/psy_data.rst:
```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+the overall checksum value.
+
+The two API-specific ReadOnlyVerification libraries are both based
+on this base class. They only implement the checksum functions
```
Done.

hiker commented on examples/gocean/eg7/test.x90:
```diff
@@ -45,13 +45,19 @@ Program test
     call parallel_init()
+    call read_only_verify_PSyDataInit()
+    ! Just to show how to use the start function. It is not
+    ! actually required, since the previous call enables it anyway.
```
Ticket opened (and note that I think just changing the default at startup might not fully be what you want: if your code explicitly does PSyDataStart() somewhere, you likely still want to disable it ... or? See #889 for further discussion).

hiker commented on lib/Makefile:
```diff
@@ -0,0 +1,48 @@
+#------------------------------------------------------------------------------
+# BSD 2-Clause License
```
Done (in base class and all readonly libs Makefiles).

hiker commented on lib/Makefile:
```diff
@@ -0,0 +1,48 @@
+#------------------------------------------------------------------------------
+# Authors: J. Henrichs, Australian Bureau of Meteorology
+
+# This Makefile just creates the Fortran base class from the
```
Done.

hiker commented on lib/Makefile:
```diff
@@ -0,0 +1,48 @@
+F90 ?= gfortran
+F90FLAGS ?= 
+
+default: psy_data_base.o
```
Yes (and several other Makefiles as well).

hiker commented on lib/process.py:
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+
+'''This modules reads a jinja template file, processes it
```
Done.

arporter replied:
You still have "modules" at line 3? Having written that, I now realise this file is missing license/copyright info too.

hiker commented on lib/process.py:
```diff
@@ -0,0 +1,70 @@
+
+parser = argparse.ArgumentParser(
+    description='Process a jinja template for PSyData.')
+parser.add_argument('template_name', help="Name of the template to process.")
```
Done.

hiker commented on lib/process.py:
```diff
@@ -0,0 +1,70 @@
+parser = argparse.ArgumentParser(
+    description='Process a jinja template for PSyData.')
+parser.add_argument('template_name', help="Name of the template to process.")
+parser.add_argument('-types', help="Comma-separated list of types, "
```
Both done.

hiker commented on lib/process.py:
```diff
@@ -0,0 +1,70 @@
+                                  "e.g. 1,2,4.",
+                    default="1,2,3,4")
+
+parser.add_argument("-prefix", help="Prefix for PSyData functions.",
```
Done.

hiker commented on lib/process.py:
```diff
@@ -0,0 +1,70 @@
+for my_type in types:
+    if my_type not in TYPE_DATA:
+        print("Type '{0}' is not supported.".format(my_type))
+        print("Use one or more of {0}"
```
OK, all done.

hiker commented on lib/read_only/Makefile:
```diff
@@ -0,0 +1,56 @@
+#------------------------------------------------------------------------------
+# BSD 2-Clause License
```
Done.

hiker commented on lib/read_only/Makefile:
```diff
@@ -0,0 +1,56 @@
+F90 ?= gfortran
+F90FLAGS ?= 
+
+PROCESS_ARGS = -prefix=read_only_verify_ -types=int,real,double \
```
Done.

hiker commented on lib/read_only/dl_esm_inf/Makefile:
```diff
@@ -44,21 +44,33 @@ INF_DIR ?= $(SHARED_DIR)/dl_esm_inf/finite_difference
 %.o: %.f90
 	$(F90) $(F90FLAGS) -c $<
 
+psy_data_base.f90: ../../psy_data_base.jinja
```
Done.

arporter replied:
Great, thanks.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -33,56 +33,29 @@
-!> This module implements a verification that read-only fields (in the LFRic
-!! infrastructure) are not overwritten (due to memory overwrites etc).
+!> This module implements a verification that read-only fields (in the
+!! dl_esm_inf infrastructure) are not overwritten (due to memory overwrites etc).
```
First one is done. Second one isn't since this is not a jinja created file. I've added a description of using jinja or not for the actual libraries to the documentation as well to make this difference more obvious. And I've also removed all intermediate files (i.e. jinja output files) from the repo (as agreed upon in our telco).

hiker commented on lib/read_only/lfric/Makefile:
```diff
@@ -42,25 +42,40 @@
+
+read_only.o: read_only_base.o
+
+%.f90:	%.jinja
```
Done.

hiker commented on lib/read_only/lfric/Makefile:
```diff
@@ -42,25 +42,40 @@
 %.o: %.f90
 	$(F90) $(F90FLAGS) -c $<
 
+psy_data_base.f90: ../../psy_data_base.jinja
```
Done.

hiker commented on lib/read_only/lfric/read_only.jinja:
```diff
@@ -0,0 +1,225 @@
+{{ "! DO NOT MODIFY DIRECTLY                            " }}
+{{ "! ==================================================" }}
+
+{% if ALL_DIMS is not defined %}
```
Done.

arporter replied:
Great, thanks.

hiker commented on lib/read_only/lfric/read_only.jinja:
```diff
@@ -0,0 +1,225 @@
+   {% set ALL_DIMS = [1, 2, 3, 4] %}
+{% endif %}
+
+{# The types that are supported. The first entry of each tuple
```
Done.

hiker commented:
I've reworked the documentation quite a bit, I hope it's easier to understand now. I also fixed a bug (dl_esm_inf did not export the functions for 2d-real arrays, which would be used when a corresponding grid property is used by a kernel. I've also added this to the example, so we kind of have a test case).

hiker commented:
Codevoc seems happy, so ready to go.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -602,20 +602,39 @@ Jinja Support:
+Jinja is used in the generic base class ``PSyDataBase``, and the base class for all
+``ReadOnly`` libraries. It is not required that any library using one of these base
+classes uses itself Jinja. For example, the ReadOnly library for
+dl_esm_inf is not using Jinja (except for processing the base class templates
```
s/is not using/does not use/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -602,20 +602,39 @@ Jinja Support:
+
+Jinja is used in the generic base class ``PSyDataBase``, and the base class for all
+``ReadOnly`` libraries. It is not required that any library using one of these base
+classes uses itself Jinja. For example, the ReadOnly library for
```
s/uses itself/itself uses/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -602,20 +602,39 @@ Jinja Support:
+implementation of ``PreDeclareVariable`` must be provided for each data type.
+For LFRic that can easily result in over 10 very similar subroutines (3 basic
+types integer, 4- and 8-byte reals; and 4- and 8-byte arrays of one to four
+dimensions). In order to simplify the creation of these subroutine the templating
```
s/subroutine/subroutines/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -602,20 +602,39 @@ Jinja Support:
+For LFRic that can easily result in over 10 very similar subroutines (3 basic
+types integer, 4- and 8-byte reals; and 4- and 8-byte arrays of one to four
+dimensions). In order to simplify the creation of these subroutine the templating
+language Jinja is being used. Jinja allows to create code based on an template,
```
s/allows to create/creates/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -655,23 +674,28 @@ verification library (taken from ``lib/read_only/lfric/Makefile``):
+
+Below is a short excerpt that shows how these variables are defined by
+default, and how they are used to create subroutines and declare their
+parameters in ``lib//read_only_base.jinja``:
```
Did you intend the double slash '//' here?

hiker replied:
No :)

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -710,9 +734,25 @@ read-only-verification base class needs to be adjusted as well.
-The ``FieldDouble`` and ``FieldVectorDouble`` related functions are explicitly
-coded in the library, the rest is taken from the base class:
+
+Since the PSyData relies on a generic interface to automatically call the
```
s/PSyData relies/PSyData API relies/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -710,9 +734,25 @@ read-only-verification base class needs to be adjusted as well.
+
+Since the PSyData relies on a generic interface to automatically call the
+right subroutine depending on type, a library must declare these automatically
+created subroutine (together with additional, library-specific versions)
```
s/subroutine/subroutines/

hiker replied:
Done.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -710,9 +734,25 @@ read-only-verification base class needs to be adjusted as well.
+right subroutine depending on type, a library must declare these automatically
+created subroutine (together with additional, library-specific versions)
+as one generic interface. This can either be done by explicitly listing the
+subroutines (which is for example done in ``lib/read_only/dl_esm_inf/read_onlyf90``,
```
"read_only.f90"

hiker replied:
Indeed.

arporter commented on doc/developer_guide/psy_data.rst:
```diff
@@ -743,10 +783,10 @@ coded in the library, the rest is taken from the base class:
-.. note:
-    Using ``or ""`` in the jinja statements avoids that ``None`` is
+.. note::
+    Using ``or ""`` in the Jinja statements avoids having ``None``
```
The "``or" bit of this note is not rendering correctly for me (pdf).

hiker replied:
It looks correct for me, though harder to see in PDF than in html. I am not sure how to improve that. I could use single quotes? Italic? None of which would be really easy to see :( Put it on a separate line??

arporter commented on lib/read_only/Makefile:
```diff
@@ -1,30 +1,36 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019, Australian Bureau of Meteorology
```
2020?

hiker replied:
Done.

arporter commented on lib/read_only/dl_esm_inf/Makefile:
```diff
@@ -1,30 +1,36 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019, Australian Bureau of Meteorology
```
2020?

hiker replied:
Done.

arporter commented on lib/read_only/lfric/Makefile:
```diff
@@ -1,30 +1,36 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019, Australian Bureau of Meteorology
```
2020?

hiker replied:
Done.

arporter commented on lib/read_only/lfric/Makefile:
```diff
@@ -42,8 +48,15 @@
+# The LFRic ReadOnly library needs the PreDeclareVariable() and
+# ProvideVariable() routines for the basic Fortran types real, int,
+# and double, and for 1-, 2-, 3-, and 4-dimensional arrays. These arguments
+# are used for # both base classes: PSyDataBase and ReadOnlyBase to ensure
```
s/for # both/for both/ I think?

hiker replied:
Yep.

arporter reviewed (changes requested):
I've not yet built the code as I attempted to pip install 'jinja' and pip crashed! That made me realise that we need to update the developer-guide installation instructions and probably setup.py too.
Therefore, I'm passing this back now.
The updated documentation is now much better and all of the additional comments have really helped too.

hiker commented on lib/Makefile:
```diff
@@ -1,35 +1,41 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019, Australian Bureau of Meteorology
```
Indeed.

hiker commented:
I've added a psyapi-section to setup.py, and updated the system-specific documentation to show how to use this section. I am not entirely sure if this is the best way, but since for now we have no way of officially distributing the wrapper libraries, I think it should be ok for now.

hiker commented:
FWIW, here is the PDF view of the ``or ""``.
![orquote](https://user-images.githubusercontent.com/6138652/93180010-da12f380-f779-11ea-9678-be38fddf3be1.png)
Admittedly not ideal, but it's there. Let me know if it's not visible for you at all. I suggested a few things that might improve this a bit in a comment further up.

Otherwise ready for next review.

arporter commented:
Thanks @hiker. I think I must be missing something. On my machine, that fragment you sent looks like:
![Screenshot from 2020-09-15 09-29-55](https://user-images.githubusercontent.com/17409212/93186106-24728100-f736-11ea-96be-f6349b7e4f3f.png)
As you can see, there's nothing before the "or" and I'm not sure what should be there?

hiker commented:
> 
> 
> Thanks @hiker. I think I must be missing something. On my machine, that fragment you sent looks like:
> ![Screenshot from 2020-09-15 09-29-55](https://user-images.githubusercontent.com/17409212/93186106-24728100-f736-11ea-96be-f6349b7e4f3f.png)
> As you can see, there's nothing before the "or" and I'm not sure what should be there?

Looks like my email reply did not make it into git. So here again:

That is (I think) what it is supposed to look like. It is marked as 'code' (``), which means in latex it is set as a fix-width font (courier I'd guess), same as the None a few words later. It's easier to see in html (where 'code' has a different background).

arporter commented:
Aha! I knew I was being stupid. How about "...ending the Jinja statements with 'or ""' avoids..." i.e. put single quotes around it. (I was reading it as "... with <missing> or "" avoids ..." - I didn't realise the Jinja code was 'or ""'.

arporter commented on lib/process.py:
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
```
Attempting to do `make` in `lib/read_only/dl_esm_inf` fails because this script is not executable. It also failes if dl_esm_inf has not already been compiled. Unless there's a good reason not to, please could you extend the Makefile to build the library too if necessary.

hiker replied:
OK, missing 'x' for process.py ... thanks to my windows file system (fwiw, I am going to switch to a native linux FS on my new VM, which has more disk space.) I'll add the dl_esm_inf compilation.

Your final compilation problem is then a left-over from missing execute permissions of process.py: it created an empty file, which was then compiled, creating ... nothing :) If you remove read_only_base.f90 (or do a ``make clean``) it should work. But I'll start working on the other bits and pieces tomorrow.

arporter replied:
Aha! I'd guessed it was your windows file system but didn't think to do a `make clean`. I've done that and now have built the dl_esm_inf read-only wrapper successfully. While I'm here, is there an example that demonstrates this?

hiker replied:
examples/gocean/eg7
I am having now some compilation problems there ... the created alg.f90 file contains >132 characters, so we either need to add ``-l output`` to psyclone, or the flag to gfortran to accept lines >132 characters (-ffree-line-length-none I believe).

arporter replied:
I'd go with the PSyclone solution if possible - try and minimise compiler-specific dependencies.

hiker replied:
Done.

arporter reviewed (changes requested):
All of my requested changes (so far) have been addressed.
I've followed the new instructions (thanks) and have Jinja2 installed.
However, when I proceed to try and build I hit problems:

1. The `lib/process.py` script is not executable.
2. If I fix 1. and try going to read_only/dl_esm_inf and doing Make then it complains that the base modules aren't available.
3. If I go to lib and do `make` and then to read_only and do `make` and then to dl_esm_inf and do `make`, I still get:

    gfortran  -I../../../external/dl_esm_inf/finite_difference/src -c read_only.f90
    read_only.f90:47:8:

   47 |     use read_only_base_mod, only : ReadOnlyBaseType, read_only_verify_PSyDataInit, &
      |        1
    Fatal Error: Cannot open module file ‘read_only_base_mod.mod’ for reading at (1): No such file or  directory

Finally, I've realised the lib/README.md needs updating and the various sub-dirs need READMEs adding. (Possibly, once you've done that I'll be able to see where I'm going wrong. :-) )

hiker commented:
> Aha! I knew I was being stupid. How about "...ending the Jinja statements with 'or ""' avoids..." i.e. put single quotes around it. (I was reading it as "... with _missing-value_ or "" avoids ..." - I didn't realise the Jinja code was 'or ""'.

OK, done. Now it looks maybe a bit odd in html, but I think it's ok. I added single quotes around other special characters in that paragraph to be consistent.

hiker commented:
> However, when I proceed to try and build I hit problems:
> 
>     1. The `lib/process.py` script is not executable.

Fixed.

>     2. If I fix 1. and try going to read_only/dl_esm_inf and doing Make then it complains that the base modules aren't available.

It should now compile this automatically (at this stage it might trigger a recompile of some files in dl_esm_inf. I have local patches for that, but am waiting for Sergi's work to be merged first).

>     3. If I go to lib and do `make` and then to read_only and do `make` and then to dl_esm_inf and do `make`, I still get:
>        gfortran  -I../../../external/dl_esm_inf/finite_difference/src -c read_only.f90
>        read_only.f90:47:8:
>        47 |     use read_only_base_mod, only : ReadOnlyBaseType, read_only_verify_PSyDataInit, &
>        |        1
>        Fatal Error: Cannot open module file ‘read_only_base_mod.mod’ for reading at (1): No such file or  directory

As discussed, this requires a ``make clean`` (even if you update your branch, if you still have an empty file you might need to manually clean it).

> Finally, I've realised the lib/README.md needs updating and the various sub-dirs need READMEs adding. (Possibly, once you've done that I'll be able to see where I'm going wrong. :-) )

Oh gee - yes, READMEs were on my todo list, and completely forgot. I've added/updated README to ``lib``, ``lib/read_only``, and ``lib/read_only/{dl_esm_inf,lfric}``.

hiker commented:
And added yet-another-readme to the gocean example.
This should be ready for another review now.

arporter commented on examples/gocean/eg7/README.md:
```diff
@@ -0,0 +1,92 @@
+
+## Compilation
+A makefile is provided to compile this example. If required
+it will compile the dl_esm_inf libraryW and the read-only verification
```
spurious 'W'

hiker replied:
How on earth did my editor not flag this ... it is supposed to do syntax checking. Guess I need to figure out how to not ignore camel-case or so. Anyway, fixed.

arporter commented on examples/gocean/eg7/README.md:
```diff
@@ -0,0 +1,92 @@
+within the PSyclone repository. You can set the environment variable
+``INF_DIR`` for the ``make`` command to pick a different version.
+
+The makefile here will invoke psyclone with the ``--profile invokes``
```
Cut-n-paste error? It uses the `read_only_transform.py` transformation script.

hiker replied:
Indeed.

arporter commented on examples/gocean/eg7/README.md:
```diff
@@ -0,0 +1,92 @@
+The makefile here will invoke psyclone with the ``--profile invokes``
+flag, which will add profiling around the two invokes used in the example.
+
+The source files overwrites the two read-only variables
```
s/files/code/

hiker replied:
OK.

arporter commented on examples/gocean/eg7/README.md:
```diff
@@ -0,0 +1,92 @@
+flag, which will add profiling around the two invokes used in the example.
+
+The source files overwrites the two read-only variables
+``b_fld`` and ``z`` in the kernel called, even though these variables are
```
s/called//

hiker replied:
Done.

arporter commented on lib/README.md:
```diff
@@ -1,9 +1,33 @@
+## `read_only` directory
+
+Contains PSyData wrapper libraries for verifying at run time that parameters declared
+as read-only in the PSyclone meta data are indeed not changed in a subroutine.
```
s/meta data/metadata/

hiker replied:
Done.

arporter commented on lib/read_only/README.md:
```diff
@@ -0,0 +1,29 @@
+# ReadOnly Verification Libraries
+
+This directories contains files related to read-only verification, i.e. checks at
```
s/directories/directory/ ?

hiker replied:
Indeed.

arporter commented on lib/read_only/dl_esm_inf/README.md:
```diff
@@ -0,0 +1,29 @@
+# Read-only Verification Library for dl_esm_info
```
s/info/inf/ :-)

hiker replied:
Arg - we have to fix the naming :cry:

arporter reviewed (changes requested):
Builds out of the box for me now.
Many thanks for adding all the READMEs.
There are just a few typos/cut-n-paste errors and then this is ready to go.

hiker commented:
Ready to go.

arporter commented:
All good now. I've taken the liberty of correcting an over-zealous camel-casing of 'PSyclone' in the README. Once Travis is happy I'll merge this.
